doc/aaa/AAA Improvements/AAA Improvements.md

```diff
@@ -0,0 +1,636 @@
+
+**STEP 4)** Next, the sshd monitor calls **`pam_setcred()`**. According to the *PAM Application Developers' Guide*, this API "*establishes, maintains, and/or deletes the credentials of a user*." The sshd monitor can thus set additional user credentials after successful authentication and before a session is opened with `pam_open_session()`. The Developers' Guide defines a credential as "*something that the user possesses, such as **supplementary group** membership*." The Guide clarifies that "*on a Linux system, the user's **UID** and **GID** are credentials too, but these properties are credentials that should be set directly by the application and not by PAM. Such credentials should be established, by the application, before a call to this function*", which is precisely what the sshd monitor does by invoking `getpwnam()` to retrieve the user's **UID** and **GID** before calling any PAM APIs.
+
+**STEP 5)** The API **`pam_open_session()`** marks the beginning of the user session. This API and its counterpart, **`pam_close_session()`**, are typically used for accounting purposes. For example, they can be used to record the start, end, and duration of a session in a logging system such as the syslog.
+
+**STEP 6)** After all PAM operations complete successfully, the sshd monitor forks itself one last time. The forked child drops its privileges to the user's UID and GID, and the supplementary groups are retrieved from `/etc/group` and applied to the child with the help of the POSIX API [**`initgroups()`**](http://man7.org/linux/man-pages/man3/initgroups.3.html). This API initializes the group access list by reading the group database `/etc/group` and using all groups of which the user is a member. Supplementary groups that the user was added to earlier by `pam_setcred()` are thus applied to the current process.
+
+**STEP 7)** Finally, the child process forks one last time to execute the `shell` program associated with the user. The shell program runs with the same UID, Primary GID, and supplementary GIDs of the process that forked it.
+
+## Current SONiC AAA implementation and limitations
+
+- **Lack of support for Multiple Roles**. The current RADIUS/TACACS+ PAM/NSS modules save the user's role, received from the RADIUS/TACACS+ server to the Primary GID associated with that user, in `/etc/passwd`. The current design assumes that the RADIUS/TACACS+ server only assigns a single role to users. That is not the case, especially in the context of [Role-Based Access Control (RBAC)](https://en.wikipedia.org/wiki/Role-based_access_control). RBAC is a way to configure Authorization policies for users. With RBAC, we assign one or more roles to users (e.g., system admin, security admin, network admin, network operator, ...). Using the Primary GID does not work since a single GID cannot represent multiple roles. 
```

Comment:
What is the target sonic customer deployments trying to address here? 

Reply:
Could you clarify the question? Are you asking me to describe RBAC?